R/teal_data_utils.R

```diff
@@ -0,0 +1,84 @@
+#' `teal_data` utils
+#'
+#' In `teal` we need to recreate the `teal_data` object due to two operations:integer.max
+#' - we need to append filter-data code and objects which have been evaluated in `FilteredData` and
+#' we want to avoid double-evaluation.
+#' - we need to subset `teal_data` to datanames used by the module, to shorten obtainable R-code
+#'
+#' Due to above recreation of `teal_data` object can't be done simply by using public
+#' `teal.code` and `teal.data` methods.
+#'
+#' @param data (`teal_data`)
+#' @param code (`character`) code to append to `data@code`
+#' @param objects (`list`) objects to append to `data@env`
+#' @param datanames (`character`) names of the datasets
+#' @return modified `teal_data`
+#' @name teal_data_utilities
+NULL
+
+#' @rdname teal_data_utilities
+.append_evaluated_code <- function(data, code) {
+  checkmate::assert_class(data, "teal_data")
+  data@code <- c(data@code, code)
+  data@id <- c(data@id, sample.int(.Machine$integer.max, size = length(code)))
+  data@messages <- c(data@messages, rep("", length(code)))
+  data@warnings <- c(data@warnings, rep("", length(code)))
+  methods::validObject(data)
+  data
+}
+
+#' @rdname teal_data_utilities
+.append_modified_data <- function(data, objects) {
+  checkmate::assert_class(data, "teal_data")
+  checkmate::assert_class(objects, "list")
+  new_env <- list2env(objects, parent = .GlobalEnv)
+  rlang::env_coalesce(new_env, data@env)
+  data@env <- new_env
+  data
+}
+
+#' @rdname teal_data_utilities
+.subset_teal_data <- function(data, datanames) {
+  checkmate::assert_class(data, "teal_data")
+  checkmate::assert_class(datanames, "character")
+  datanames_corrected <- intersect(datanames, ls(data@env))
+  dataname_corrected_with_raw <- intersect(c(datanames, sprintf("%s_raw", datanames)), ls(data@env))
+
+  if (!length(datanames)) {
+    return(teal_data())
+  }
+
+  new_data <- do.call(
+    teal.data::teal_data,
+    args = c(
+      mget(x = dataname_corrected_with_raw, envir = data@env),
+      list(
+        code = gsub(
+          "warning('Code was not verified for reproducibility.')\n",
+          "",
+          teal.data::get_code(data, datanames = dataname_corrected_with_raw),
+          fixed = TRUE
+        ),
+        join_keys = teal.data::join_keys(data)[datanames_corrected]
+      )
+    )
+  )
+  new_data@verified <- data@verified
+  teal.data::datanames(new_data) <- datanames_corrected
+  new_data
+}
+
+#' @rdname teal_data_utilities
+.teal_data_datanames <- function(data) {
+  checkmate::assert_class(data, "teal_data")
+  if (length(teal.data::datanames(data))) {
+    teal.data::datanames(data)
+  } else {
+    .teal_data_ls(data)
+  }
+}
+
+#' @rdname teal_data_utilities
+.teal_data_ls <- function(data) {
+  ls(teal.code::get_env(data), all.names = TRUE)
```

Comment:
@gogonzo I think you need to resolve some conflicts with `669_insertUI@main` because those lines that got merged into the feature branch are not on this branch - https://github.com/insightsengineering/teal/pull/1305/files#diff-60bef5af43a1bd541299ce9bceeb226afb23deaa5e637b145c85e05ffd552b0cR209-R210

This happened because you changed the name of the function, and with this branch you are appending new function, and not modyfing the old one, that already got updated by @vedhav

Reply:
Yes, I'm working on it - not easy to fix it now. Please give me some time

Reply:
@vedhav I think `_raw` datasets provided in `data` should be ignored with the message (or at least if there are `dataset` and `dataset_raw` in the same time).
The failing test you've fixed afterwards kind of says that it is wrong:
When initial data is `iris` and `iris_raw`, then in the filter panel we have a problem:

```r
data <- teal_data(iris = iris, mtcars = mtcars, swiss = swiss, iris_raw = iris)
# in the filter panel
iris_raw <- iris
iris_raw_raw <- iris_raw
```

Maybe we should have more uncommon suffix instead of `_raw` - maybe `._raw_` or something?

Reply:
Ah! true. I see the problem. I'll change this to `._raw_` or see if there is some alternative way to solve this. I see that these raw datasets are only used while building the data summary. Perhaps they can be fetched without searching for datasets with the name `*_raw`.

Reply:
>  I see that these raw datasets are only used while building the data summary

`_raw` datasets are also included to the module's `data@env` but they are not included in `datanames(data)`

Reply:
I've made necessary fixes on this branch